**Autor:** Joan Campaner Gutiérrez

**Link al repositorio:**  https://github.com/Joancg98/IAFIFA

# Práctica FIFA

Se quiere predecir el *valor* de los jugadores utilizando la información sobre ellos.

## Importación de librerias

Primero de todo importamos las librerías que vamos a utilizar.

In [1]:
import os
import math

from sklearn.model_selection import train_test_split
from sklearn import linear_model
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn import preprocessing

import pandas as pd
import numpy as np


## Leer los datos

Para leer los datos, se va a utilizar ``pandas`` y una función de la librería ``os``.

In [2]:
df = pd.read_csv(os.path.join("..", "in", "fifa.csv"))
name = df["Name"]
pd.set_option('display.max_columns', None) # Permite visualizar todas las columnas
pd.options.mode.chained_assignment = None # Desactiva los warnings cuando se trabaja con slices de dataframes
df.head()

Unnamed: 0      ID               Name  Age  \
0           0  158023           L. Messi   31   
1           1   20801  Cristiano Ronaldo   33   
2           2  190871          Neymar Jr   26   
3           3  193080             De Gea   27   
4           4  192985       K. De Bruyne   27   

                                            Photo Nationality  \
0  https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1   https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2  https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3  https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4  https://cdn.sofifa.org/players/4/19/192985.png     Belgium   

                                  Flag  Overall  Potential  \
0  https://cdn.sofifa.org/flags/52.png       94         94   
1  https://cdn.sofifa.org/flags/38.png       94         94   
2  https://cdn.sofifa.org/flags/54.png       92         93   
3  https://cdn.sofifa.org/flags/45.png       91         93   
4   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club                                     Club Logo    Value  \
0         FC Barcelona  https://cdn.sofifa.org/teams/2/light/241.png  €110.5M   
1             Juventus   https://cdn.sofifa.org/teams/2/light/45.png     €77M   
2  Paris Saint-Germain   https://cdn.sofifa.org/teams/2/light/73.png  €118.5M   
3    Manchester United   https://cdn.sofifa.org/teams/2/light/11.png     €72M   
4      Manchester City   https://cdn.sofifa.org/teams/2/light/10.png    €102M   

    Wage  Special Preferred Foot  International Reputation  Weak Foot  \
0  €565K     2202           Left                       5.0        4.0   
1  €405K     2228          Right                       5.0        4.0   
2  €290K     2143          Right                       5.0        5.0   
3  €260K     1471          Right                       4.0        3.0   
4  €355K     2281          Right                       4.0        5.0   

   Skill Moves       Work Rate   Body Type Real Face Position  Jersey Number  \
0          4.0  Medium/ Medium       Messi       Yes       RF           10.0   
1          5.0       High/ Low  C. Ronaldo       Yes       ST            7.0   
2          5.0    High/ Medium      Neymar       Yes       LW           10.0   
3          1.0  Medium/ Medium        Lean       Yes       GK            1.0   
4          4.0      High/ High      Normal       Yes      RCM            7.0   

         Joined Loaned From Contract Valid Until Height  Weight    LS    ST  \
0   Jul 1, 2004         NaN                 2021    5'7  159lbs  88+2  88+2   
1  Jul 10, 2018         NaN                 2022    6'2  183lbs  91+3  91+3   
2   Aug 3, 2017         NaN                 2022    5'9  150lbs  84+3  84+3   
3   Jul 1, 2011         NaN                 2020    6'4  168lbs   NaN   NaN   
4  Aug 30, 2015         NaN                 2023   5'11  154lbs  82+3  82+3   

     RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   

    RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     8

## Analisis de datos

Primero de todo, queremos saber qué está ocurriendo a los datos. Para hacerlo, vamos a incrementar el número de columnas que se ven. Después de eso, utilizaremos una instrucción especial que nos enseñará la estructura de datos

In [3]:
df.describe()

Unnamed: 0             ID           Age       Overall     Potential  \
count  18207.000000   18207.000000  18207.000000  18207.000000  18207.000000   
mean    9103.000000  214298.338606     25.122206     66.238699     71.307299   
std     5256.052511   29965.244204      4.669943      6.908930      6.136496   
min        0.000000      16.000000     16.000000     46.000000     48.000000   
25%     4551.500000  200315.500000     21.000000     62.000000     67.000000   
50%     9103.000000  221759.000000     25.000000     66.000000     71.000000   
75%    13654.500000  236529.500000     28.000000     71.000000     75.000000   
max    18206.000000  246620.000000     45.000000     94.000000     95.000000   

            Special  International Reputation     Weak Foot   Skill Moves  \
count  18207.000000              18159.000000  18159.000000  18159.000000   
mean    1597.809908                  1.113222      2.947299      2.361308   
std      272.586016                  0.394031      0.660456      0.756164   
min      731.000000                  1.000000      1.000000      1.000000   
25%     1457.000000                  1.000000      3.000000      2.000000   
50%     1635.000000                  1.000000      3.000000      2.000000   
75%     1787.000000                  1.000000      3.000000      3.000000   
max     2346.000000                  5.000000      5.000000      5.000000   

       Jersey Number      Crossing     Finishing  HeadingAccuracy  \
count   18147.000000  18159.000000  18159.000000     18159.000000   
mean       19.546096     49.734181     45.550911        52.298144   
std        15.947765     18.364524     19.525820        17.379909   
min         1.000000      5.000000      2.000000         4.000000   
25%         8.000000     38.000000     30.000000        44.000000   
50%        17.000000     54.000000     49.000000        56.000000   
75%        26.000000     64.000000     62.000000        64.000000   
max        99.000000     93.000000     95.000000        94.000000   

       ShortPassing       Volleys     Dribbling         Curve    FKAccuracy  \
count  18159.000000  18159.000000  18159.000000  18159.000000  18159.000000   
mean      58.686712     42.909026     55.371001     47.170824     42.863153   
std       14.699495     17.694408     18.910371     18.395264     17.478763   
min        7.000000      4.000000      4.000000      6.000000      3.000000   
25%       54.000000     30.000000     49.000000     34.000000     31.000000   
50%       62.000000     44.000000     61.000000     48.000000     41.000000   
75%       68.000000     57.000000     68.000000     62.000000     57.000000   
max       93.000000     90.000000     97.000000     94.000000     94.000000   

        LongPassing   BallControl  Acceleration   SprintSpeed       Agility  \
count  18159.000000  18159.000000  18159.000000  18159.000000  18159.000000   
mean      52.711933     58.369459     64.614076     64.726967     63.503607   
std       15.327870     16.686595     14.927780     14.649953     14.766049   
min        9.000000      5.000000     12.000000     12.000000     14.000000   
25%       43.000000     54.000000     57.000000     57.000000     55.000000   
50%       56.000000     63.000000     67.000000     67.000000     66.000000   
75%       64.000000     69.000000     75.000000     75.000000     74.000000   
max       93.000000     96.000000     97.000000     96.000000     96.000000   

          Reactions       Balance     ShotPower       Jumping       Stamina  \
count  18159.000000  18159.000000  18159.000000  18159.000000  18159.000000   
mean      61.836610     63.966573     55.460047     65.089432     63.219946   
std        9.010464     14.136166     17.237958     11.820044     15.894741   
min       21.000000     16.000000      2.000000     15.000000     12.000000   
25%       56.000000     56.000000     45.000000     58.000000     56.000000   
50%       62.000000     66.000000     59.000000     66.000000     66.00000

Podemos ver que las dos primeras columnas no aportan ninguna información relevante.
Procedemos a eliminarlas

In [4]:
df = df.iloc[:, 2:]
df

Name  Age  \
0                L. Messi   31   
1       Cristiano Ronaldo   33   
2               Neymar Jr   26   
3                  De Gea   27   
4            K. De Bruyne   27   
...                   ...  ...   
18202        J. Lundstram   19   
18203  N. Christoffersson   19   
18204           B. Worman   16   
18205      D. Walker-Rice   17   
18206           G. Nugent   16   

                                                Photo Nationality  \
0      https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1       https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2      https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3      https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4      https://cdn.sofifa.org/players/4/19/192985.png     Belgium   
...                                               ...         ...   
18202  https://cdn.sofifa.org/players/4/19/238813.png     England   
18203  https://cdn.sofifa.org/players/4/19/243165.png      Sweden   
18204  https://cdn.sofifa.org/players/4/19/241638.png     England   
18205  https://cdn.sofifa.org/players/4/19/246268.png     England   
18206  https://cdn.sofifa.org/players/4/19/246269.png     England   

                                      Flag  Overall  Potential  \
0      https://cdn.sofifa.org/flags/52.png       94         94   
1      https://cdn.sofifa.org/flags/38.png       94         94   
2      https://cdn.sofifa.org/flags/54.png       92         93   
3      https://cdn.sofifa.org/flags/45.png       91         93   
4       https://cdn.sofifa.org/flags/7.png       91         92   
...                                    ...      ...        ...   
18202  https://cdn.sofifa.org/flags/14.png       47         65   
18203  https://cdn.sofifa.org/flags/46.png       47         63   
18204  https://cdn.sofifa.org/flags/14.png       47         67   
18205  https://cdn.sofifa.org/flags/14.png       47         66   
18206  https://cdn.sofifa.org/flags/14.png       46         66   

                      Club                                       Club Logo  \
0             FC Barcelona    https://cdn.sofifa.org/teams/2/light/241.png   
1                 Juventus     https://cdn.sofifa.org/teams/2/light/45.png   
2      Paris Saint-Germain     https://cdn.sofifa.org/teams/2/light/73.png   
3        Manchester United     https://cdn.sofifa.org/teams/2/light/11.png   
4          Manchester City     https://cdn.sofifa.org/teams/2/light/10.png   
...                    ...                                             ...   
18202      Crewe Alexandra    https://cdn.sofifa.org/teams/2/light/121.png   
18203       Trelleborgs FF    https://cdn.sofifa.org/teams/2/light/703.png   
18204     Cambridge United   https://cdn.sofifa.org/teams/2/light/1944.png   
18205      Tranmere Rovers  https://cdn.sofifa.org/teams/2/light/15048.png   
18206      Tranmere Rovers  https://cdn.sofifa.org/teams/2/light/15048.png   

         Value   Wage  Special Preferred Foot  International Reputation  \
0      €110.5M  €565K     2202           Left                       5.0   
1         €77M  €405K     2228          Right                       5.0   
2      €118.5M  €290K     2143          Right                       5.0   
3         €72M  €260K     1471          Right                       4.0   
4        €102M  €355K     2281          Right                       4.0   
...        ...    ...      ...            ...                       ...   
18202     €60K    €1K     1307          Right                       1.0   
18203     €60K    €1K     1098          Right                       1.0   
18204     €60K    €1K     1189          Right                       1.0   
18205     €60K    €1K     1228          Right                       1.0   
18206     €60K    €1K     1321          Right                       1.0   

       Weak Foot  Skill Moves       Work Rate   Body Type Real Face Position  \
0            4.0          4.0  Medium/ Medium       Messi       Yes     

Borramos todas las columnas posiciones y sus modificadores. Estas columnas son parecidas a las estadísticas del jugador y son algo difíciles de manejar, por eso se opta por borrarlas.

In [5]:
positions_to_drop = []
for i in range(26, 52):
   positions_to_drop.append(df.columns[i])
df = df.drop(positions_to_drop, axis = 1)
df

Name  Age  \
0                L. Messi   31   
1       Cristiano Ronaldo   33   
2               Neymar Jr   26   
3                  De Gea   27   
4            K. De Bruyne   27   
...                   ...  ...   
18202        J. Lundstram   19   
18203  N. Christoffersson   19   
18204           B. Worman   16   
18205      D. Walker-Rice   17   
18206           G. Nugent   16   

                                                Photo Nationality  \
0      https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1       https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2      https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3      https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4      https://cdn.sofifa.org/players/4/19/192985.png     Belgium   
...                                               ...         ...   
18202  https://cdn.sofifa.org/players/4/19/238813.png     England   
18203  https://cdn.sofifa.org/players/4/19/243165.png      Sweden   
18204  https://cdn.sofifa.org/players/4/19/241638.png     England   
18205  https://cdn.sofifa.org/players/4/19/246268.png     England   
18206  https://cdn.sofifa.org/players/4/19/246269.png     England   

                                      Flag  Overall  Potential  \
0      https://cdn.sofifa.org/flags/52.png       94         94   
1      https://cdn.sofifa.org/flags/38.png       94         94   
2      https://cdn.sofifa.org/flags/54.png       92         93   
3      https://cdn.sofifa.org/flags/45.png       91         93   
4       https://cdn.sofifa.org/flags/7.png       91         92   
...                                    ...      ...        ...   
18202  https://cdn.sofifa.org/flags/14.png       47         65   
18203  https://cdn.sofifa.org/flags/46.png       47         63   
18204  https://cdn.sofifa.org/flags/14.png       47         67   
18205  https://cdn.sofifa.org/flags/14.png       47         66   
18206  https://cdn.sofifa.org/flags/14.png       46         66   

                      Club                                       Club Logo  \
0             FC Barcelona    https://cdn.sofifa.org/teams/2/light/241.png   
1                 Juventus     https://cdn.sofifa.org/teams/2/light/45.png   
2      Paris Saint-Germain     https://cdn.sofifa.org/teams/2/light/73.png   
3        Manchester United     https://cdn.sofifa.org/teams/2/light/11.png   
4          Manchester City     https://cdn.sofifa.org/teams/2/light/10.png   
...                    ...                                             ...   
18202      Crewe Alexandra    https://cdn.sofifa.org/teams/2/light/121.png   
18203       Trelleborgs FF    https://cdn.sofifa.org/teams/2/light/703.png   
18204     Cambridge United   https://cdn.sofifa.org/teams/2/light/1944.png   
18205      Tranmere Rovers  https://cdn.sofifa.org/teams/2/light/15048.png   
18206      Tranmere Rovers  https://cdn.sofifa.org/teams/2/light/15048.png   

         Value   Wage  Special Preferred Foot  International Reputation  \
0      €110.5M  €565K     2202           Left                       5.0   
1         €77M  €405K     2228          Right                       5.0   
2      €118.5M  €290K     2143          Right                       5.0   
3         €72M  €260K     1471          Right                       4.0   
4        €102M  €355K     2281          Right                       4.0   
...        ...    ...      ...            ...                       ...   
18202     €60K    €1K     1307          Right                       1.0   
18203     €60K    €1K     1098          Right                       1.0   
18204     €60K    €1K     1189          Right                       1.0   
18205     €60K    €1K     1228          Right                       1.0   
18206     €60K    €1K     1321          Right                       1.0   

       Weak Foot  Skill Moves       Work Rate   Body Type Real Face Position  \
0            4.0          4.0  Medium/ Medium       Messi       Yes     

Procedemos a eliminar otras columnas que no aportan información como ``Photo`` o ``Club Logo``.

In [6]:
# Flag, Photo, Club Logo: Se eliminan por ser imágenes
# Work Rate: se elimina por ser una estadística varible con el tiempo
# Body Type, Jersey Number, Real Face: Se eliminan por ser características sin valor
# Joined: Se elimina porque se considera que el valor no dependerá de cuándo se unió al club

# Nationality: Se ha probado a considerar esta característica igual que con Club, pero al hacerlo, el modelo aumenta
# en muchas columnas y deja de funcionar correctamente. Se deduce que el valor no depende de dónde ha nacido el jugador.
df = df.drop(['Flag','Photo', 'Body Type', 'Jersey Number', 'Real Face', 'Work Rate', 'Club Logo', 'Joined'], axis=1)
df = df.drop(['Loaned From', 'Nationality'], axis=1)
df

Name  Age  Overall  Potential                 Club  \
0                L. Messi   31       94         94         FC Barcelona   
1       Cristiano Ronaldo   33       94         94             Juventus   
2               Neymar Jr   26       92         93  Paris Saint-Germain   
3                  De Gea   27       91         93    Manchester United   
4            K. De Bruyne   27       91         92      Manchester City   
...                   ...  ...      ...        ...                  ...   
18202        J. Lundstram   19       47         65      Crewe Alexandra   
18203  N. Christoffersson   19       47         63       Trelleborgs FF   
18204           B. Worman   16       47         67     Cambridge United   
18205      D. Walker-Rice   17       47         66      Tranmere Rovers   
18206           G. Nugent   16       46         66      Tranmere Rovers   

         Value   Wage  Special Preferred Foot  International Reputation  \
0      €110.5M  €565K     2202           Left                       5.0   
1         €77M  €405K     2228          Right                       5.0   
2      €118.5M  €290K     2143          Right                       5.0   
3         €72M  €260K     1471          Right                       4.0   
4        €102M  €355K     2281          Right                       4.0   
...        ...    ...      ...            ...                       ...   
18202     €60K    €1K     1307          Right                       1.0   
18203     €60K    €1K     1098          Right                       1.0   
18204     €60K    €1K     1189          Right                       1.0   
18205     €60K    €1K     1228          Right                       1.0   
18206     €60K    €1K     1321          Right                       1.0   

       Weak Foot  Skill Moves Position Contract Valid Until Height  Weight  \
0            4.0          4.0       RF                 2021    5'7  159lbs   
1            4.0          5.0       ST                 2022    6'2  183lbs   
2            5.0          5.0       LW                 2022    5'9  150lbs   
3            3.0          1.0       GK                 2020    6'4  168lbs   
4            5.0          4.0      RCM                 2023   5'11  154lbs   
...          ...          ...      ...                  ...    ...     ...   
18202        2.0          2.0       CM                 2019    5'9  134lbs   
18203        2.0          2.0       ST                 2020    6'3  170lbs   
18204        3.0          2.0       ST                 2021    5'8  148lbs   
18205        3.0          2.0       RW                 2019   5'10  154lbs   
18206        3.0          2.0       CM                 2019   5'10  176lbs   

       Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0          84.0       95.0             70.0          90.0     86.0       97.0   
1          84.0       94.0             89.0          81.0     87.0       88.0   
2          79.0       87.0             62.0          84.0     84.0       96.0   
3          17.0       13.0             21.0          50.0     13.0       18.0   
4          93.0       82.0             55.0          92.0     82.0       86.0   
...         ...        ...              ...           ...      ...        ...   
18202      34.0       38.0             40.0          49.0     25.0       42.0   
18203      23.0       52.0             52.0          43.0     36.0       39.0   
18204      25.0       40.0             46.0          38.0     38.0       45.0   
18205      44.0       50.0             39.0          42.0     40.0       51.0   
18206      41.0       34.0             46.0          48.0     30.0       43.0   

       Curve  FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  \
0       93.0        94.0         87.0         96.0          91.0         86.0   
1       81.0        76.0         77.0         94.0          89.0         91.0   
2       88.0        87.0         78.0         95.0          94.0  

## Solucionar NaNs

Qué NaNs tenemos?

In [7]:
df.columns[df.isna().any()].tolist()

['Club',
 'Preferred Foot',
 'International Reputation',
 'Weak Foot',
 'Skill Moves',
 'Position',
 'Contract Valid Until',
 'Height',
 'Weight',
 'Crossing',
 'Finishing',
 'HeadingAccuracy',
 'ShortPassing',
 'Volleys',
 'Dribbling',
 'Curve',
 'FKAccuracy',
 'LongPassing',
 'BallControl',
 'Acceleration',
 'SprintSpeed',
 'Agility',
 'Reactions',
 'Balance',
 'ShotPower',
 'Jumping',
 'Stamina',
 'Strength',
 'LongShots',
 'Aggression',
 'Interceptions',
 'Positioning',
 'Vision',
 'Penalties',
 'Composure',
 'Marking',
 'StandingTackle',
 'SlidingTackle',
 'GKDiving',
 'GKHandling',
 'GKKicking',
 'GKPositioning',
 'GKReflexes',
 'Release Clause']

In [8]:
df[df.isna().any(axis=1)]

Name  Age  Overall  Potential  \
28         J. Rodríguez   26       88         89   
38           G. Higuaín   30       88         88   
91             Paulinho   29       85         85   
166    Anderson Talisca   24       83         90   
176          M. Kovačić   24       83         89   
...                 ...  ...      ...        ...   
17979        Zhang Yuan   21       51         60   
18026           Chen Ji   21       50         59   
18031          Zhou Xin   20       50         61   
18056         L. Valeau   19       50         65   
18183     K. Pilkington   44       48         48   

                                 Club   Value   Wage  Special Preferred Foot  \
28                  FC Bayern München  €69.5M  €315K     2171           Left   
38                              Milan    €57M  €245K     1965          Right   
91     Guangzhou Evergrande Taobao FC    €37M  €235K     2245          Right   
166    Guangzhou Evergrande Taobao FC  €36.5M   €18K     2142           Left   
176                           Chelsea    €35M  €165K     2115          Right   
...                               ...     ...    ...      ...            ...   
17979             Guizhou Hengfeng FC    €60K    €2K     1459          Right   
18026             Guizhou Hengfeng FC    €50K    €2K     1440          Right   
18031                  Stabæk Fotball    €40K    €2K     1278          Right   
18056                          Ascoli    €60K    €3K     1450           Left   
18183                Cambridge United      €0    €1K      774          Right   

       International Reputation  Weak Foot  Skill Moves Position  \
28                          4.0        3.0          4.0      LAM   
38                          4.0        4.0          3.0       LS   
91                          3.0        3.0          3.0      LDM   
166                         2.0        4.0          4.0      CAM   
176                         3.0        4.0          4.0      LCM   
...                         ...        ...          ...      ...   
17979                       1.0        2.0          2.0       CM   
18026                       1.0        2.0          2.0       LM   
18031                       1.0        3.0          2.0       RB   
18056                       1.0        3.0          2.0       CM   
18183                       1.0        2.0          1.0       GK   

      Contract Valid Until Height  Weight  Crossing  Finishing  \
28            Jun 30, 2019   5'11  172lbs      90.0       83.0   
38            Jun 30, 2019    6'1  196lbs      68.0       92.0   
91            Jun 30, 2019    6'0  157lbs      64.0       81.0   
166           Dec 31, 2018    6'3  176lbs      78.0       80.0   
176           Jun 30, 2019    5'9  172lbs      73.0       62.0   
...                    ...    ...     ...       ...        ...   
17979         Dec 31, 2018    6'0  172lbs      36.0       30.0   
18026         Dec 31, 2018   5'10  143lbs      41.0       28.0   
18031         Dec 31, 2018    5'7  143lbs      46.0       29.0   
18056         Jun 30, 2019   5'10  143lbs      50.0       30.0   
18183                 2023    6'2  190lbs      11.0       12.0   

       HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  \
28                62.0          89.0     90.0       85.0   89.0        86.0   
38                80.0          75.0     90.0       84.0   74.0        62.0   
91                85.0          84.0     75.0       82.0   66.0        72.0   
166               77.0          81.0     76.0       82.0   86.0        87.0   
176               42.0          87.0     76.0       88.0   78.0        68.0   
...                ...           ...      ...        ...    ...         ...   
17979             48.0          57.0     32.0       45.0   37.0        38.0   
18026             50.0          54.0     26.0       50.0   39.0        40.0   
18031             35.0          32.0     26.0       52.0   32.0        28.0   
18056             50.0          5

Observamos que casi todos los jugadores tienen algún NaN. Podemos ver algunas columnas que son casi siempre Nan como "Loaned From". Podemos obtener todos los jugadores con la columna **Club** igual a NaN. 

In [9]:
df.loc[df.Club != df.Club]

Name  Age  Overall  Potential Club Value Wage  Special  \
452        L. Paredes   24       80         85  NaN    €0   €0     2122   
538      A. Granqvist   33       80         80  NaN    €0   €0     1797   
568          A. Lunev   26       79         81  NaN    €0   €0     1217   
677      I. Smolnikov   29       79         79  NaN    €0   €0     2038   
874         A. Dzyuba   29       78         78  NaN    €0   €0     1810   
...               ...  ...      ...        ...  ...   ...  ...      ...   
17197        V. Kaith   21       55         64  NaN    €0   €0      838   
17215        K. Taimi   26       55         57  NaN    €0   €0     1366   
17339        S. Passi   23       54         63  NaN    €0   €0     1321   
17436  D. Lalhlimpuia   20       54         67  NaN    €0   €0     1270   
17539        C. Singh   21       53         62  NaN    €0   €0     1247   

      Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
452            Right                       2.0        4.0          4.0   
538            Right                       2.0        4.0          2.0   
568            Right                       1.0        3.0          1.0   
677            Right                       2.0        3.0          3.0   
874            Right                       2.0        3.0          3.0   
...              ...                       ...        ...          ...   
17197          Right                       1.0        2.0          1.0   
17215          Right                       1.0        3.0          2.0   
17339          Right                       1.0        3.0          2.0   
17436          Right                       1.0        3.0          2.0   
17539          Right                       1.0        3.0          2.0   

      Position Contract Valid Until Height  Weight  Crossing  Finishing  \
452         CM                  NaN   5'11  165lbs      76.0       55.0   
538        LCB                  NaN    6'4  185lbs      49.0       51.0   
568         GK                  NaN    6'2  176lbs      16.0       14.0   
677         RB                  NaN   5'10  154lbs      73.0       61.0   
874         ST                  NaN    6'5  201lbs      61.0       79.0   
...        ...                  ...    ...     ...       ...        ...   
17197       GK                  NaN    6'2  176lbs      13.0        7.0   
17215       RB                  NaN    6'4  187lbs      42.0       41.0   
17339      NaN                  NaN    5'9  143lbs      35.0       56.0   
17436      NaN                  NaN    6'0  168lbs      25.0       65.0   
17539      NaN                  NaN    6'3  174lbs      22.0       21.0   

       HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  \
452               60.0          84.0     73.0       78.0   79.0        78.0   
538               81.0          73.0     37.0       49.0   36.0        40.0   
568               17.0          25.0     13.0       15.0   18.0        17.0   
677               69.0          79.0     57.0       72.0   49.0        46.0   
874               86.0          71.0     74.0       71.0   64.0        60.0   
...                ...           ...      ...        ...    ...         ...   
17197             13.0          16.0      7.0       11.0   10.0        13.0   
17215             65.0          45.0     26.0       50.0   30.0        32.0   
17339             49.0          38.0     38.0       53.0   37.0        33.0   
17436             48.0          38.0     48.0       47.0   33.0        30.0   
17539             48.0          36.0     20.0       38.0   21.0        27.0   

       LongPassing  BallControl  Acceleration  SprintSpeed  Agility  \
452           82.0         82.0          75.0         69.0     77.0   
538           67.0         63.0          46.0         49.0     55.0   
568           32.0         17.0          58.0         54.0     36.0   
677           75.0         72.0          84.0         90.0     80.0   
874           55.0    

El problema de estos jugadores es que su club no pertenece a la FIFA. Como podemos ver, son solo 241 jugadores por lo que podemos borrarlos del dataset.

In [10]:
df = df.dropna(subset = ['Club'])
df

Name  Age  Overall  Potential                 Club  \
0                L. Messi   31       94         94         FC Barcelona   
1       Cristiano Ronaldo   33       94         94             Juventus   
2               Neymar Jr   26       92         93  Paris Saint-Germain   
3                  De Gea   27       91         93    Manchester United   
4            K. De Bruyne   27       91         92      Manchester City   
...                   ...  ...      ...        ...                  ...   
18202        J. Lundstram   19       47         65      Crewe Alexandra   
18203  N. Christoffersson   19       47         63       Trelleborgs FF   
18204           B. Worman   16       47         67     Cambridge United   
18205      D. Walker-Rice   17       47         66      Tranmere Rovers   
18206           G. Nugent   16       46         66      Tranmere Rovers   

         Value   Wage  Special Preferred Foot  International Reputation  \
0      €110.5M  €565K     2202           Left                       5.0   
1         €77M  €405K     2228          Right                       5.0   
2      €118.5M  €290K     2143          Right                       5.0   
3         €72M  €260K     1471          Right                       4.0   
4        €102M  €355K     2281          Right                       4.0   
...        ...    ...      ...            ...                       ...   
18202     €60K    €1K     1307          Right                       1.0   
18203     €60K    €1K     1098          Right                       1.0   
18204     €60K    €1K     1189          Right                       1.0   
18205     €60K    €1K     1228          Right                       1.0   
18206     €60K    €1K     1321          Right                       1.0   

       Weak Foot  Skill Moves Position Contract Valid Until Height  Weight  \
0            4.0          4.0       RF                 2021    5'7  159lbs   
1            4.0          5.0       ST                 2022    6'2  183lbs   
2            5.0          5.0       LW                 2022    5'9  150lbs   
3            3.0          1.0       GK                 2020    6'4  168lbs   
4            5.0          4.0      RCM                 2023   5'11  154lbs   
...          ...          ...      ...                  ...    ...     ...   
18202        2.0          2.0       CM                 2019    5'9  134lbs   
18203        2.0          2.0       ST                 2020    6'3  170lbs   
18204        3.0          2.0       ST                 2021    5'8  148lbs   
18205        3.0          2.0       RW                 2019   5'10  154lbs   
18206        3.0          2.0       CM                 2019   5'10  176lbs   

       Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0          84.0       95.0             70.0          90.0     86.0       97.0   
1          84.0       94.0             89.0          81.0     87.0       88.0   
2          79.0       87.0             62.0          84.0     84.0       96.0   
3          17.0       13.0             21.0          50.0     13.0       18.0   
4          93.0       82.0             55.0          92.0     82.0       86.0   
...         ...        ...              ...           ...      ...        ...   
18202      34.0       38.0             40.0          49.0     25.0       42.0   
18203      23.0       52.0             52.0          43.0     36.0       39.0   
18204      25.0       40.0             46.0          38.0     38.0       45.0   
18205      44.0       50.0             39.0          42.0     40.0       51.0   
18206      41.0       34.0             46.0          48.0     30.0       43.0   

       Curve  FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  \
0       93.0        94.0         87.0         96.0          91.0         86.0   
1       81.0        76.0         77.0         94.0          89.0         91.0   
2       88.0        87.0         78.0         95.0          94.0  

Seguimos viendo si queda algún NaN en nuestro dataset

In [11]:
df.columns[df.isna().any()].tolist()

['Preferred Foot',
 'International Reputation',
 'Weak Foot',
 'Skill Moves',
 'Position',
 'Contract Valid Until',
 'Height',
 'Weight',
 'Crossing',
 'Finishing',
 'HeadingAccuracy',
 'ShortPassing',
 'Volleys',
 'Dribbling',
 'Curve',
 'FKAccuracy',
 'LongPassing',
 'BallControl',
 'Acceleration',
 'SprintSpeed',
 'Agility',
 'Reactions',
 'Balance',
 'ShotPower',
 'Jumping',
 'Stamina',
 'Strength',
 'LongShots',
 'Aggression',
 'Interceptions',
 'Positioning',
 'Vision',
 'Penalties',
 'Composure',
 'Marking',
 'StandingTackle',
 'SlidingTackle',
 'GKDiving',
 'GKHandling',
 'GKKicking',
 'GKPositioning',
 'GKReflexes',
 'Release Clause']

Podemos ver que todavía quedan. Pasamos a ver los jugadores que no tienen una elección de pie preferido

In [12]:
df.loc[df['Preferred Foot'] != df['Preferred Foot']]

Name  Age  Overall  Potential                           Club  \
13236      J. McNulty   33       62         62                       Rochdale   
13237      J. Barrera   29       62         62                Boyacá Chicó FC   
13238        J. Stead   35       62         62                   Notts County   
13239     A. Semprini   20       62         72                        Brescia   
13240      R. Bingham   24       62         66         Hamilton Academical FC   
13241    K. Dankowski   21       62         72                  Śląsk Wrocław   
13242       I. Colman   23       62         70         Club Atlético Aldosivi   
13243       M. Feeney   19       62         78                        Everton   
13244        R. Minor   30       62         62                       Hobro IK   
13245          Klauss   21       62         69                   HJK Helsinki   
13246      I. Sissoko   22       62         68                     AS Béziers   
13247         F. Hart   28       62         62                 SV Mattersburg   
13248   L. McCullough   24       62         69                Tranmere Rovers   
13249       Li Yunqiu   27       62         62  Shanghai Greenland Shenhua FC   
13250       F. Garcia   29       62         62               Itagüí Leones FC   
13251    R. Haemhouts   34       62         62                      NAC Breda   
13252       E. Binaku   22       62         70                       Malmö FF   
13253       G. Miller   31       62         62                Carlisle United   
13254      A. Aidonis   17       62         82                  VfB Stuttgart   
13255        L. Sowah   25       62         65         Hamilton Academical FC   
13256       R. Deacon   26       62         62                      Dundee FC   
13257   Jang Hyun Soo   25       62         65        Suwon Samsung Bluewings   
13258     A. Al Malki   23       62         67                       Al Wehda   
13259     E. Guerrero   27       62         65                   CD Palestino   
13260         Hernáiz   20       62         69                    Albacete BP   
13261   H. Al Mansour   25       62         64                       Al Nassr   
13262         H. Paul   24       62         66               TSV 1860 München   
13263        S. Bauer   25       62         66       FC Admira Wacker Mödling   
13264      M. Chergui   29       62         62               Grenoble Foot 38   
13265      D. Gardner   28       62         62                Oldham Athletic   
13266    L. Bengtsson   20       62         73                    Hammarby IF   
13267    F. Jaramillo   22       62         70               Itagüí Leones FC   
13268      L. Garguła   37       62         62                  Miedź Legnica   
13269       S. Rivera   26       62         64            Jaguares de Córdoba   
13270        Vinicius   19       62         77                        Bologna   
13271    F. Sepúlveda   26       62         63                 CD Antofagasta   
13272       L. Spence   22       62         70                      Dundee FC   
13273      B. Lepistu   25       62         67                Kristiansund BK   
13274     A. Abruscia   27       62         62               TSV 1860 München   
13275     E. González   23       62         70                Boyacá Chicó FC   
13276      M. Al Amri   26       62         63                        Al Raed   
13277    J. Rebolledo   26       62         62               Deportes Iquique   
13278      C. Mamengi   17       62         79                     FC Utrecht   
13279    P. Mazzocchi   22       62         69                        Perugia   
13280       Y. Ammour   19       62         77                Montpellier HSC   
13281  Jwa Joon Hyeop   27       62         62                   Gyeongnam FC   
13282      O. Marrufo   25       62         65    Tiburones Rojos de Veracruz   
13283     Han Pengfei   25       62         66            Guizhou Hengfeng FC   

       Value Wage  Special Preferred Foot  Inte

Se observa que de todos estos jugadores nos faltan bastantes estadísticas. Como tampoco son tantos, se decide borrarlos.

In [13]:
df = df.dropna(subset = ['Preferred Foot'])
df

Name  Age  Overall  Potential                 Club  \
0                L. Messi   31       94         94         FC Barcelona   
1       Cristiano Ronaldo   33       94         94             Juventus   
2               Neymar Jr   26       92         93  Paris Saint-Germain   
3                  De Gea   27       91         93    Manchester United   
4            K. De Bruyne   27       91         92      Manchester City   
...                   ...  ...      ...        ...                  ...   
18202        J. Lundstram   19       47         65      Crewe Alexandra   
18203  N. Christoffersson   19       47         63       Trelleborgs FF   
18204           B. Worman   16       47         67     Cambridge United   
18205      D. Walker-Rice   17       47         66      Tranmere Rovers   
18206           G. Nugent   16       46         66      Tranmere Rovers   

         Value   Wage  Special Preferred Foot  International Reputation  \
0      €110.5M  €565K     2202           Left                       5.0   
1         €77M  €405K     2228          Right                       5.0   
2      €118.5M  €290K     2143          Right                       5.0   
3         €72M  €260K     1471          Right                       4.0   
4        €102M  €355K     2281          Right                       4.0   
...        ...    ...      ...            ...                       ...   
18202     €60K    €1K     1307          Right                       1.0   
18203     €60K    €1K     1098          Right                       1.0   
18204     €60K    €1K     1189          Right                       1.0   
18205     €60K    €1K     1228          Right                       1.0   
18206     €60K    €1K     1321          Right                       1.0   

       Weak Foot  Skill Moves Position Contract Valid Until Height  Weight  \
0            4.0          4.0       RF                 2021    5'7  159lbs   
1            4.0          5.0       ST                 2022    6'2  183lbs   
2            5.0          5.0       LW                 2022    5'9  150lbs   
3            3.0          1.0       GK                 2020    6'4  168lbs   
4            5.0          4.0      RCM                 2023   5'11  154lbs   
...          ...          ...      ...                  ...    ...     ...   
18202        2.0          2.0       CM                 2019    5'9  134lbs   
18203        2.0          2.0       ST                 2020    6'3  170lbs   
18204        3.0          2.0       ST                 2021    5'8  148lbs   
18205        3.0          2.0       RW                 2019   5'10  154lbs   
18206        3.0          2.0       CM                 2019   5'10  176lbs   

       Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0          84.0       95.0             70.0          90.0     86.0       97.0   
1          84.0       94.0             89.0          81.0     87.0       88.0   
2          79.0       87.0             62.0          84.0     84.0       96.0   
3          17.0       13.0             21.0          50.0     13.0       18.0   
4          93.0       82.0             55.0          92.0     82.0       86.0   
...         ...        ...              ...           ...      ...        ...   
18202      34.0       38.0             40.0          49.0     25.0       42.0   
18203      23.0       52.0             52.0          43.0     36.0       39.0   
18204      25.0       40.0             46.0          38.0     38.0       45.0   
18205      44.0       50.0             39.0          42.0     40.0       51.0   
18206      41.0       34.0             46.0          48.0     30.0       43.0   

       Curve  FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  \
0       93.0        94.0         87.0         96.0          91.0         86.0   
1       81.0        76.0         77.0         94.0          89.0         91.0   
2       88.0        87.0         78.0         95.0          94.0  

In [14]:
df.columns[df.isna().any()].tolist()

['Release Clause']

In [15]:
df.loc[df['Release Clause'] != df['Release Clause']]

Name  Age  Overall  Potential  \
28         J. Rodríguez   26       88         89   
38           G. Higuaín   30       88         88   
91             Paulinho   29       85         85   
166    Anderson Talisca   24       83         90   
176          M. Kovačić   24       83         89   
...                 ...  ...      ...        ...   
17979        Zhang Yuan   21       51         60   
18026           Chen Ji   21       50         59   
18031          Zhou Xin   20       50         61   
18056         L. Valeau   19       50         65   
18183     K. Pilkington   44       48         48   

                                 Club   Value   Wage  Special Preferred Foot  \
28                  FC Bayern München  €69.5M  €315K     2171           Left   
38                              Milan    €57M  €245K     1965          Right   
91     Guangzhou Evergrande Taobao FC    €37M  €235K     2245          Right   
166    Guangzhou Evergrande Taobao FC  €36.5M   €18K     2142           Left   
176                           Chelsea    €35M  €165K     2115          Right   
...                               ...     ...    ...      ...            ...   
17979             Guizhou Hengfeng FC    €60K    €2K     1459          Right   
18026             Guizhou Hengfeng FC    €50K    €2K     1440          Right   
18031                  Stabæk Fotball    €40K    €2K     1278          Right   
18056                          Ascoli    €60K    €3K     1450           Left   
18183                Cambridge United      €0    €1K      774          Right   

       International Reputation  Weak Foot  Skill Moves Position  \
28                          4.0        3.0          4.0      LAM   
38                          4.0        4.0          3.0       LS   
91                          3.0        3.0          3.0      LDM   
166                         2.0        4.0          4.0      CAM   
176                         3.0        4.0          4.0      LCM   
...                         ...        ...          ...      ...   
17979                       1.0        2.0          2.0       CM   
18026                       1.0        2.0          2.0       LM   
18031                       1.0        3.0          2.0       RB   
18056                       1.0        3.0          2.0       CM   
18183                       1.0        2.0          1.0       GK   

      Contract Valid Until Height  Weight  Crossing  Finishing  \
28            Jun 30, 2019   5'11  172lbs      90.0       83.0   
38            Jun 30, 2019    6'1  196lbs      68.0       92.0   
91            Jun 30, 2019    6'0  157lbs      64.0       81.0   
166           Dec 31, 2018    6'3  176lbs      78.0       80.0   
176           Jun 30, 2019    5'9  172lbs      73.0       62.0   
...                    ...    ...     ...       ...        ...   
17979         Dec 31, 2018    6'0  172lbs      36.0       30.0   
18026         Dec 31, 2018   5'10  143lbs      41.0       28.0   
18031         Dec 31, 2018    5'7  143lbs      46.0       29.0   
18056         Jun 30, 2019   5'10  143lbs      50.0       30.0   
18183                 2023    6'2  190lbs      11.0       12.0   

       HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  \
28                62.0          89.0     90.0       85.0   89.0        86.0   
38                80.0          75.0     90.0       84.0   74.0        62.0   
91                85.0          84.0     75.0       82.0   66.0        72.0   
166               77.0          81.0     76.0       82.0   86.0        87.0   
176               42.0          87.0     76.0       88.0   78.0        68.0   
...                ...           ...      ...        ...    ...         ...   
17979             48.0          57.0     32.0       45.0   37.0        38.0   
18026             50.0          54.0     26.0       50.0   39.0        40.0   
18031             35.0          32.0     26.0       52.0   32.0        28.0   
18056             50.0          5

Los siguientes jugadores no tienen una cláusula de rescinsión. Podemos borrarlos también ya que no es una gran cantidad del dataset.

In [16]:
df = df.dropna(subset = ['Release Clause'])

In [17]:
df.columns[df.isna().any()].tolist()

[]

Observamos que ya no tenemos más NaN en nuestro dataset. 

Pasamos a convertir los datos como ``Value`` o ``Wage`` a numéricos.

In [18]:
def value_to_float(x):
    """
    From K and M to float.
    
    """
    x = x.replace('€', '')
    ret_val = 0.0
    
    if type(x) == float or type(x) == int:
        ret_val = x
    if 'K' in x:
        if len(x) > 1:
            ret_val = float(x.replace('K', ''))
        ret_val = ret_val *1000
    if 'M' in x:
        if len(x) > 1:
            ret_val = float(x.replace('M', ''))
        ret_val = ret_val * 1000000.0
    return ret_val

Aplicamos la función a cada valor de la columna

In [19]:
df.loc[:,('Value')] = df.loc[:,('Value')].apply(value_to_float)
df.loc[:,('Wage')] = df.loc[:,('Wage')].apply(value_to_float)
df.loc[:,('Release Clause')] = df.loc[:,('Release Clause')].apply(value_to_float)

# Convertimos la fecha de finalización de contrato en un valor float.
df1 = df.pop('Contract Valid Until') # remove column b and store it in df1
df1 = pd.to_numeric(df1, downcast = "float")
df['Contract Valid Until'] = df1 # add b series as a 'new' column.
df

Name  Age  Overall  Potential                 Club  \
0                L. Messi   31       94         94         FC Barcelona   
1       Cristiano Ronaldo   33       94         94             Juventus   
2               Neymar Jr   26       92         93  Paris Saint-Germain   
3                  De Gea   27       91         93    Manchester United   
4            K. De Bruyne   27       91         92      Manchester City   
...                   ...  ...      ...        ...                  ...   
18202        J. Lundstram   19       47         65      Crewe Alexandra   
18203  N. Christoffersson   19       47         63       Trelleborgs FF   
18204           B. Worman   16       47         67     Cambridge United   
18205      D. Walker-Rice   17       47         66      Tranmere Rovers   
18206           G. Nugent   16       46         66      Tranmere Rovers   

             Value      Wage  Special Preferred Foot  \
0      110500000.0  565000.0     2202           Left   
1       77000000.0  405000.0     2228          Right   
2      118500000.0  290000.0     2143          Right   
3       72000000.0  260000.0     1471          Right   
4      102000000.0  355000.0     2281          Right   
...            ...       ...      ...            ...   
18202      60000.0    1000.0     1307          Right   
18203      60000.0    1000.0     1098          Right   
18204      60000.0    1000.0     1189          Right   
18205      60000.0    1000.0     1228          Right   
18206      60000.0    1000.0     1321          Right   

       International Reputation  Weak Foot  Skill Moves Position Height  \
0                           5.0        4.0          4.0       RF    5'7   
1                           5.0        4.0          5.0       ST    6'2   
2                           5.0        5.0          5.0       LW    5'9   
3                           4.0        3.0          1.0       GK    6'4   
4                           4.0        5.0          4.0      RCM   5'11   
...                         ...        ...          ...      ...    ...   
18202                       1.0        2.0          2.0       CM    5'9   
18203                       1.0        2.0          2.0       ST    6'3   
18204                       1.0        3.0          2.0       ST    5'8   
18205                       1.0        3.0          2.0       RW   5'10   
18206                       1.0        3.0          2.0       CM   5'10   

       Weight  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0      159lbs      84.0       95.0             70.0          90.0     86.0   
1      183lbs      84.0       94.0             89.0          81.0     87.0   
2      150lbs      79.0       87.0             62.0          84.0     84.0   
3      168lbs      17.0       13.0             21.0          50.0     13.0   
4      154lbs      93.0       82.0             55.0          92.0     82.0   
...       ...       ...        ...              ...           ...      ...   
18202  134lbs      34.0       38.0             40.0          49.0     25.0   
18203  170lbs      23.0       52.0             52.0          43.0     36.0   
18204  148lbs      25.0       40.0             46.0          38.0     38.0   
18205  154lbs      44.0       50.0             39.0          42.0     40.0   
18206  176lbs      41.0       34.0             46.0          48.0     30.0   

       Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0           97.0   93.0        94.0         87.0         96.0          91.0   
1           88.0   81.0        76.0         77.0         94.0          89.0   
2           96.0   88.0        87.0         78.0         95.0          94.0   
3           18.0   21.0        19.0         51.0         42.0          57.0   
4           86.0   85.0        83.0         91.0         91.0          78.0   
...          ...    ...         ...          ...          ...           ...   
18202       42.0   30.0        34.0         45.0        

Pasamos a convertir los datos de peso y altura a valores float. 

In [20]:
def weight_to_float(x):
    
    x = x.replace('lbs', '')
   
    return float(x)

In [21]:
def height_to_float(height):
    height = height.split("'")
    feet = float(height[0])
    inch = float(height[1])

    inch += feet * 12
    cm = round(inch * 2.54, 1)
    return cm

In [22]:
df.loc[:,('Weight')] = df.loc[:,('Weight')].apply(weight_to_float)
df.loc[:,('Height')] = df.loc[:,('Height')].apply(height_to_float)
df

Name  Age  Overall  Potential                 Club  \
0                L. Messi   31       94         94         FC Barcelona   
1       Cristiano Ronaldo   33       94         94             Juventus   
2               Neymar Jr   26       92         93  Paris Saint-Germain   
3                  De Gea   27       91         93    Manchester United   
4            K. De Bruyne   27       91         92      Manchester City   
...                   ...  ...      ...        ...                  ...   
18202        J. Lundstram   19       47         65      Crewe Alexandra   
18203  N. Christoffersson   19       47         63       Trelleborgs FF   
18204           B. Worman   16       47         67     Cambridge United   
18205      D. Walker-Rice   17       47         66      Tranmere Rovers   
18206           G. Nugent   16       46         66      Tranmere Rovers   

             Value      Wage  Special Preferred Foot  \
0      110500000.0  565000.0     2202           Left   
1       77000000.0  405000.0     2228          Right   
2      118500000.0  290000.0     2143          Right   
3       72000000.0  260000.0     1471          Right   
4      102000000.0  355000.0     2281          Right   
...            ...       ...      ...            ...   
18202      60000.0    1000.0     1307          Right   
18203      60000.0    1000.0     1098          Right   
18204      60000.0    1000.0     1189          Right   
18205      60000.0    1000.0     1228          Right   
18206      60000.0    1000.0     1321          Right   

       International Reputation  Weak Foot  Skill Moves Position  Height  \
0                           5.0        4.0          4.0       RF   170.2   
1                           5.0        4.0          5.0       ST   188.0   
2                           5.0        5.0          5.0       LW   175.3   
3                           4.0        3.0          1.0       GK   193.0   
4                           4.0        5.0          4.0      RCM   180.3   
...                         ...        ...          ...      ...     ...   
18202                       1.0        2.0          2.0       CM   175.3   
18203                       1.0        2.0          2.0       ST   190.5   
18204                       1.0        3.0          2.0       ST   172.7   
18205                       1.0        3.0          2.0       RW   177.8   
18206                       1.0        3.0          2.0       CM   177.8   

       Weight  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0       159.0      84.0       95.0             70.0          90.0     86.0   
1       183.0      84.0       94.0             89.0          81.0     87.0   
2       150.0      79.0       87.0             62.0          84.0     84.0   
3       168.0      17.0       13.0             21.0          50.0     13.0   
4       154.0      93.0       82.0             55.0          92.0     82.0   
...       ...       ...        ...              ...           ...      ...   
18202   134.0      34.0       38.0             40.0          49.0     25.0   
18203   170.0      23.0       52.0             52.0          43.0     36.0   
18204   148.0      25.0       40.0             46.0          38.0     38.0   
18205   154.0      44.0       50.0             39.0          42.0     40.0   
18206   176.0      41.0       34.0             46.0          48.0     30.0   

       Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0           97.0   93.0        94.0         87.0         96.0          91.0   
1           88.0   81.0        76.0         77.0         94.0          89.0   
2           96.0   88.0        87.0         78.0         95.0          94.0   
3           18.0   21.0        19.0         51.0         42.0          57.0   
4           86.0   85.0        83.0         91.0         91.0          78.0   
...          ...    ...         ...          ...          ...           ...   
18202       42.0   30.0        34.0         

Ahora queremos convertir los valores como ``Position`` o ``Club`` a valores numéricos

In [23]:
clb = df.pop('Club')

df = pd.concat([df.reset_index(drop=True), pd.get_dummies(clb, prefix='clb').reset_index(drop=True)], axis=1, sort=False)

In [24]:
pref_foot = df.pop('Preferred Foot')

df = pd.concat([df.reset_index(drop=True), pd.get_dummies(pref_foot, prefix='pref_foot').reset_index(drop=True)], axis=1, sort=False)

In [25]:
pos = df.pop('Position')

df = pd.concat([df.reset_index(drop=True), pd.get_dummies(pos, prefix='pos').reset_index(drop=True)], axis=1, sort=False)
df

Name  Age  Overall  Potential        Value      Wage  \
0                L. Messi   31       94         94  110500000.0  565000.0   
1       Cristiano Ronaldo   33       94         94   77000000.0  405000.0   
2               Neymar Jr   26       92         93  118500000.0  290000.0   
3                  De Gea   27       91         93   72000000.0  260000.0   
4            K. De Bruyne   27       91         92  102000000.0  355000.0   
...                   ...  ...      ...        ...          ...       ...   
16638        J. Lundstram   19       47         65      60000.0    1000.0   
16639  N. Christoffersson   19       47         63      60000.0    1000.0   
16640           B. Worman   16       47         67      60000.0    1000.0   
16641      D. Walker-Rice   17       47         66      60000.0    1000.0   
16642           G. Nugent   16       46         66      60000.0    1000.0   

       Special  International Reputation  Weak Foot  Skill Moves  Height  \
0         2202                       5.0        4.0          4.0   170.2   
1         2228                       5.0        4.0          5.0   188.0   
2         2143                       5.0        5.0          5.0   175.3   
3         1471                       4.0        3.0          1.0   193.0   
4         2281                       4.0        5.0          4.0   180.3   
...        ...                       ...        ...          ...     ...   
16638     1307                       1.0        2.0          2.0   175.3   
16639     1098                       1.0        2.0          2.0   190.5   
16640     1189                       1.0        3.0          2.0   172.7   
16641     1228                       1.0        3.0          2.0   177.8   
16642     1321                       1.0        3.0          2.0   177.8   

       Weight  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0       159.0      84.0       95.0             70.0          90.0     86.0   
1       183.0      84.0       94.0             89.0          81.0     87.0   
2       150.0      79.0       87.0             62.0          84.0     84.0   
3       168.0      17.0       13.0             21.0          50.0     13.0   
4       154.0      93.0       82.0             55.0          92.0     82.0   
...       ...       ...        ...              ...           ...      ...   
16638   134.0      34.0       38.0             40.0          49.0     25.0   
16639   170.0      23.0       52.0             52.0          43.0     36.0   
16640   148.0      25.0       40.0             46.0          38.0     38.0   
16641   154.0      44.0       50.0             39.0          42.0     40.0   
16642   176.0      41.0       34.0             46.0          48.0     30.0   

       Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0           97.0   93.0        94.0         87.0         96.0          91.0   
1           88.0   81.0        76.0         77.0         94.0          89.0   
2           96.0   88.0        87.0         78.0         95.0          94.0   
3           18.0   21.0        19.0         51.0         42.0          57.0   
4           86.0   85.0        83.0         91.0         91.0          78.0   
...          ...    ...         ...          ...          ...           ...   
16638       42.0   30.0        34.0         45.0         43.0          54.0   
16639       39.0   32.0        20.0         25.0         40.0          41.0   
16640       45.0   38.0        27.0         28.0         44.0          70.0   
16641       51.0   34.0        32.0         32.0         52.0          61.0   
16642       43.0   40.0        34.0         44.0         51.0          57.0   

       SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0             86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1             91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2             90.0     96.0       94.0     84.0       80.

## Normalización
Pasamos a normalizar los datos del dataframe

In [26]:
val = df.pop('Value')

In [27]:
min_max_scaler = preprocessing.MinMaxScaler()
df.iloc[:,1:] = min_max_scaler.fit_transform(df.iloc[:, 1:])
df

Name       Age   Overall  Potential      Wage   Special  \
0                L. Messi  0.517241  1.000000   0.978723  1.000000  0.910836   
1       Cristiano Ronaldo  0.586207  1.000000   0.978723  0.716312  0.926935   
2               Neymar Jr  0.344828  0.958333   0.957447  0.512411  0.874303   
3                  De Gea  0.379310  0.937500   0.957447  0.459220  0.458204   
4            K. De Bruyne  0.379310  0.937500   0.936170  0.627660  0.959752   
...                   ...       ...       ...        ...       ...       ...   
16638        J. Lundstram  0.103448  0.020833   0.361702  0.000000  0.356656   
16639  N. Christoffersson  0.103448  0.020833   0.319149  0.000000  0.227245   
16640           B. Worman  0.000000  0.020833   0.404255  0.000000  0.283591   
16641      D. Walker-Rice  0.034483  0.020833   0.382979  0.000000  0.307740   
16642           G. Nugent  0.000000  0.000000   0.382979  0.000000  0.365325   

       International Reputation  Weak Foot  Skill Moves    Height    Weight  \
0                          1.00       0.75         0.75  0.301181  0.368421   
1                          1.00       0.75         1.00  0.651575  0.548872   
2                          1.00       1.00         1.00  0.401575  0.300752   
3                          0.75       0.50         0.00  0.750000  0.436090   
4                          0.75       1.00         0.75  0.500000  0.330827   
...                         ...        ...          ...       ...       ...   
16638                      0.00       0.25         0.25  0.401575  0.180451   
16639                      0.00       0.25         0.25  0.700787  0.451128   
16640                      0.00       0.50         0.25  0.350394  0.285714   
16641                      0.00       0.50         0.25  0.450787  0.330827   
16642                      0.00       0.50         0.25  0.450787  0.496241   

       Crossing  Finishing  HeadingAccuracy  ShortPassing   Volleys  \
0      0.897727   1.000000         0.733333      0.965116  0.953488   
1      0.897727   0.989247         0.944444      0.860465  0.965116   
2      0.840909   0.913978         0.644444      0.895349  0.930233   
3      0.136364   0.118280         0.188889      0.500000  0.104651   
4      1.000000   0.860215         0.566667      0.988372  0.906977   
...         ...        ...              ...           ...       ...   
16638  0.329545   0.387097         0.400000      0.488372  0.244186   
16639  0.204545   0.537634         0.533333      0.418605  0.372093   
16640  0.227273   0.408602         0.466667      0.360465  0.395349   
16641  0.443182   0.516129         0.388889      0.406977  0.418605   
16642  0.409091   0.344086         0.466667      0.476744  0.302326   

       Dribbling     Curve  FKAccuracy  LongPassing  BallControl  \
0       1.000000  0.988636    1.000000     0.928571     1.000000   
1       0.903226  0.852273    0.802198     0.809524     0.978022   
2       0.989247  0.931818    0.923077     0.821429     0.989011   
3       0.150538  0.170455    0.175824     0.500000     0.406593   
4       0.881720  0.897727    0.879121     0.976190     0.945055   
...          ...       ...         ...          ...          ...   
16638   0.408602  0.272727    0.340659     0.428571     0.417582   
16639   0.376344  0.295455    0.186813     0.190476     0.384615   
16640   0.440860  0.363636    0.263736     0.226190     0.428571   
16641   0.505376  0.318182    0.318681     0.273810     0.516484   
16642   0.419355  0.386364    0.340659     0.416667     0.505495   

       Acceleration  SprintSpeed   Agility  Reactions  Balance  ShotPower  \
0          0.929412     0.880952  0.939024   0.986667   0.9875   0.892473   
1          0.905882     0.940476  0.890244   1.000000   0.6750   1.000000   
2          0.964706     0.928571  1.000000   0.973333   0.8500   0.838710   
3          0.529412     0.547619  0.560976   0.920000   0.3375   0.311828   
4          0.776471     0.761905  0.792683   0.933333   

## Predicción

Finalmente queremos intentar predecir el **value**.

In [28]:
X_train, X_test, y_train, y_test = train_test_split(df.iloc[:,1:], val, test_size=0.33, random_state=42)

In [29]:
len(X_train)

11150

Ahora entrenamos el modelo de regresión lineal

In [30]:
reg = linear_model.LinearRegression()
reg.fit(X_train, y_train)
preds = reg.predict(X_test)

In [31]:
r2_score(preds, y_test)

0.9935868461974793

Obtenemos una puntuación de 0.9936, lo cual es un muy buen resultado y la predicción será bastante acertada.

In [32]:
y_test=np.array(y_test)

In [33]:
print(preds[0])
print(y_test[0])

4818432.0
5000000.0


En el ejemplo anterior observamos que un jugador con un valor de 5M de €, cuesta 4,82M de € según el modelo. Lo cual es un 96% del precio del jugador, un valor muy aproximado.

In [34]:
from sklearn.metrics import mean_absolute_error 
mean_absolute_error(y_test, preds)

213196.95903877664

El error absoluto medio es de uno 210000 €, que teniendo en cuenta los altos valores que puede tener el precio de un jugador, podemos concluir con que es un resultado bastante bueno.